contracts/FeeDisbursal/FeeDisbursal.sol

```diff
@@ -0,0 +1,269 @@
+        Reclaimable.reclaim(token);
+    }
+
+    /// @notice helper to determine if the token at the given address is supported for disbursing
+    /// @param token address of token in question
+    function isDisbursableToken(address token)
+        public
+        view
+        returns (bool)
+    {   
+        // @TODO: migrate this to new, more flexible token registering in universe
+        if (token == address(UNIVERSE.etherToken())) return true;
+        if (token == address(UNIVERSE.euroToken())) return true;
+        return false;
+    }
+
+    ////////////////////////
+    // Private functions
+    ////////////////////////
+
+    /// @notice claim a token for an spender, returns the amount of tokens claimed
+    /// @param token address of the claimable token
+    /// @param spender address of the spender that will receive the funds
+    /// @param until until what index to claim to
+    function claimPrivate(address token, address spender, uint256 until)
+    public
+    returns (uint256 claimedAmount, uint256 lastIndex)
+    {
+        require(isDisbursableToken(token), "NF_DISB_UKNOWN_TOKEN");
+        (claimedAmount, lastIndex) = claimablePrivate(token, spender, until);
+
+        // mark spender disbursal progress
+        _disbursalProgress[token][spender] = lastIndex;
+
+        // do the actual token transfer
+        IERC223Token ierc223Token = IERC223Token(token);
+        ierc223Token.transfer(spender, claimedAmount, "");
+        //@TODO: add log message
+    }
+
+    /// @notice get the amount of tokens that can be claimed by a given spender
+    /// @param token address of the claimable token
+    /// @param spender address of the spender that will receive the funds
+    /// @param until until what index to claim to, use UINT256_MAX for all
+    function claimablePrivate(address token, address spender, uint256 until)
+    public
+    view
+    returns (uint256 claimableAmount, uint256 lastIndex)
+    {
+        require(isDisbursableToken(token), "NF_DISB_UKNOWN_TOKEN");
+        lastIndex = min(until, _disbursals[token].length);
+        claimableAmount = 0;
+        for (uint256 i = _disbursalProgress[token][spender]; i < lastIndex; i += 1) {
+            Disbursal storage disbursal = _disbursals[token][i];
+            BasicSnapshotToken proRataToken = BasicSnapshotToken(disbursal.proRataToken);
+            uint256 snapshotId = disbursal.snapshotId;
+            uint256 proRataTotalSupply = proRataToken.totalSupplyAt(snapshotId);
+            uint256 proRataSpenderBalance = proRataToken.balanceOfAt(spender, snapshotId);
+            if (proRataTotalSupply == 0) continue;
+            // @TODO: do we need checks for overflow here? probably not as uint256 is really large..
+            // this should round down, so we should not be spending more than we have in our balance
+            claimableAmount += (proRataSpenderBalance * disbursal.amount) / proRataTotalSupply;
```

Comment:
please use our fully tested `Math` contract. it has `proportion` method that does exactly what you have here but for example makes sure there are no overflows and rounding is consistent.

also. we need a real proof that rounding errors here do not accumulate. if they do, the last person may not be able to claim because there is not enough balance.

but I have inductive proof that it will not happen. it's however not so obvious at all!

Reply:
Replacing the proportion part is done